### DISCUSSION TITLE
[WINDOWS] problems with variables when launching from mpv

### DISCUSSION THREAD

**Discussion by masterofobzene** [mpv-player/mpv#15589] [Q&A]
I use input.conf to launch certain cmd batch files to do stuff with files being played in mpv.

for example I use a 4 letter trigger to launch a script: 
`T-E-S-T				run "E:\\Perfil\\Desktop\\0005.TEST.c.ms.CMD" "${path}"; show-text "TEST"`

The script uses "%1" variable to deal with the file currently playing.  It seems that mpv doesn't like "%1" and just says nothing is there ONLY IF I PLAY A WHOLE FOLDER (it's contents). The variable WORKS (and the whole script too) if I only play ONE file (a single video).

To be clear; files are always played by mpv, what fails is my manually launched CMD that moves the files currently being played. 

TLDR: seems like mpv is not passing variables correctly to my batch file, in this case "%1" variable.

**Comment by masterofobzene** [comment]
I think this question will be a long trip to "we need all info" and/or "I don't understand" kind of replies so I'm going to keep posting my findings alone and if somebody knows whats going on, please reply.

Now, to focus on this:

The batch is acting weird. The file below IS NOT in the folder that it checks, but it still returns "true" as if the file was there:
```
C:\Users\PC>if exist "E:\Perfil\Desktop\[TEST_2]\2020-09-27_09-02-16_178~1__-seg37.mkv" GOTO RENAMEVID else GOTO MOVEFILEVID

C:\Users\PC>ren E:\Perfil\Desktop\TEST_1/2020-09-27_09-02-16_178~1__-seg37.mkv "2020-09-27_09-02-16_178~1__-seg37_31251.mkv"
The syntax of the command is incorrect.

C:\Users\PC>MOVE "E:\Perfil\Desktop\TEST_1\2020-09-27_09-02-16_178~1__-seg37_31251.mkv" "E:\Perfil\Desktop\[TEST_2]"
The system cannot find the file specified.
```

To dissect this:
`if exist "E:\Perfil\Desktop\[TEST_2]\2020-09-27_09-02-16_178~1__-seg37.mkv" GOTO RENAMEVID else GOTO MOVEFILEVID`

The batch is erratically thinking the file "2020-09-27_09-02-16_178~1__-seg37.mkv" is inside "E:\Perfil\Desktop\[TEST_2]" **which is not true**. The batch evaluates if the file is already in the destination folder and if its true, it renames it then moves it; if it's not true it just moves it. In this case, the batch is wrongly evaluating that the file IS in the destination. WEIRD. **At this point I want to say that this batch job works in windows directly; it only fails when launched from mpv.**

The script decides that the file must be renamed cause it "thinks" that a file with the same name is already in the destination, but pay close attention to what happened:
```
C:\Users\PC>ren E:\Perfil\Desktop\TEST_1/2020-09-27_09-02-16_178~1__-seg37.mkv "2020-09-27_09-02-16_178~1__-seg37_31251.mkv"
The syntax of the command is incorrect.
```

I will highlight it for you: 
ren E:\Perfil\Desktop\TEST_1/2020-09-27_09-02-16_178~1__-seg37.mkv
      Look for the separator here ↑

mpv for some reason is changing "%1" (in this case %1 = `E:\Perfil\Desktop\TEST_1\2020-09-27_09-02-16_178~1__-seg37.mkv`) and it's adding **/** instead of **\\** that is the windows separator.

The 3rd command obviously fails since the "renamed file" was never renamed:
```
C:\Users\PC>MOVE "E:\Perfil\Desktop\TEST_1\2020-09-27_09-02-16_178~1__-seg37_31251.mkv" "E:\Perfil\Desktop\[TEST_2]"
The system cannot find the file specified.
```

So up until here, I see many strange things:
1- I don't know why is my batch file evaluating a wrong "if exist" 
2- I don't know why MPV is adding/changing that separator to a linux one.
3- I don't know why if loading a single video into mpv, everything works.
4- I don't know why if loading many videos or a folder into mpv, the problems arise.

**Comment by masterofobzene** [comment]
It seems that when `ren E:\Perfil\Desktop\TEST_1/2020-09-27_09-02-16_178~1__-seg37.mkv` (bad separator added) it was due to a bad code in my batch:  I had `"%DESTVID%%~nx1"` instead of `"%DESTVID%\%~nx1"` 
So I added the missing "\\" and the "weird linux separator" was gone. 

One less to troubleshoot.

Sadly, the script is still evaluating wrong and insists on renaming the file:

```
if exist "E:\Perfil\Desktop\[TEST_2]\2020-09-27_09-02-16_178~1__-seg37.mkv" GOTO RENAMEVID else GOTO MOVEFILEVID

:RENAMEVID 
ren E:\Perfil\Desktop\TEST_1\2020-09-27_09-02-16_178~1__-seg37.mkv "2020-09-27_09-02-16_178~1__-seg37_9276.mkv"
```

**Comment by masterofobzene** [comment]
Well, just discovered something:

For some magical reason that I do not comprehend because my limited knowledge, "if exist" is being evaluated wrongly when launching the batch file from mpv. I assure you that this script worked without changes if used inside windows directly. So I will answer my own questions for people that end up here:

> 1- I don't know why is my batch file evaluating a wrong "if exist"


Black magic. Mpv does something when executing batch files that changes the native behaviour. The fix: using "if not exist" and flipping around the actions after it.

> 2- I don't know why MPV is adding/changing that separator to a linux one.


Auto-answered above, It was my fault in my script. Mpv was trying to "fix" my typo with "it's own separator". This is specified somewhere in the manual, but was a little vague and ambiguous. Still, it was a "typo" because I forgot to add a separator there.

> 3- I don't know why if loading a single video into mpv, everything works.


No freaking idea. Still a mistery but I suppose it has something to do with for example "%1" not working properly and having to declare it into a new variable. **This is important** "%1" argument variables (something like "current selected file") is not playing nice with mpv and needs to be re-declared into another "formal" variable e.x. `SET "SOMENAME=%1"`. **%1 is a null variable when launching batch files from mpv.** All this happens when loading multiple files, folders or playlists; _it works when launching individual files_.


I will close this because I think I solved my issue, feel free to see if this can be fixed in the future or at least answered to me here if I'm missing something.

**Reply by avih** [reply]
Normally windows applications don't have an issue with path separator being `\` or `/`, because the windows API handles them effectvely identically, and both are valid and mean the same thing.

However, one big exception is `cmd.exe` (which also interprets batch files), which doesn't use the normal windows API, and instead has its own interpreter and very weird rules (really, if you want to do yourself a favour, get away from batch files).

> 3- I don't know why if loading a single video into mpv, everything works.

Because mpv doesn't modify `${path}` if it's a direct argument to mpv, so if it was invoked with only `\` separators then the batch file will work.because `${path}` has only `\` separators.

But if the argument is a directory, then mpv generates `${path}` by concatenating the dir name and the file name - you guessed it - using `/` separator.

It does that because it's simpler code, and normally windows applications have zero issues with that, except, like you just found out, in batch files.

This was discussed quite a few times in the past, with several suggested solutions (you'll have to search the issues), some of them involve changing mpv code itself, and others using a small script which can fix it if needed.

So far the mpv developers stick to keeping it only `/` when a path is generated from components, because it is simpler code.
